extras/indexing/src/main/java/org/apache/rya/indexing/entity/storage/mongo/MongoEntityStorage.java

```diff
@@ -242,4 +283,84 @@ private static Bson makeExplicitTypeFilter(final RyaURI typeId) {
 
         return Stream.of(dataTypeFilter, valueFilter);
     }
+
+    private boolean detectDuplicates(final Entity entity) throws EntityStorageException {
+        boolean hasDuplicate = false;
+        if (duplicateDataDetector.isDetectionEnabled()) {
+            // Grab all entities that have all the same explicit types as our
+            // original Entity.
+            final List<Entity> comparisonEntities = searchHasAllExplicitTypes(entity.getExplicitTypeIds());
+
+            // Now that we have our set of potential duplicates, compare them.
+            // We can stop when we find one duplicate.
+            for (final Entity compareEntity : comparisonEntities) {
+                try {
+                    hasDuplicate = duplicateDataDetector.compareEntities(entity, compareEntity);
+                } catch (final SmartUriException e) {
+                    throw new EntityStorageException("Encountered an error while comparing entities.", e);
+                }
+                if (hasDuplicate) {
+                    break;
+                }
+            }
+        }
+        return hasDuplicate;
+    }
+
+    /**
+     * Searches the Entity storage for all Entities that contain all the
+     * specified explicit type IDs.
+     * @param explicitTypeIds the {@link ImmutableList} of {@link RyaURI}s that
+     * are being searched for.
+     * @return the {@link List} of {@link Entity}s that have all the specified
+     * explicit type IDs. If nothing was found an empty {@link List} is
+     * returned.
+     * @throws EntityStorageException
+     */
+    private List<Entity> searchHasAllExplicitTypes(final ImmutableList<RyaURI> explicitTypeIds) throws EntityStorageException {
+        final List<Entity> hasAllExplicitTypesEntities = new ArrayList<>();
+        if (!explicitTypeIds.isEmpty()) {
+            // Grab the first type from the explicit type IDs.
+            final RyaURI firstType = explicitTypeIds.get(0);
+
+            // Check if that type exists anywhere in storage.
+            final List<RyaURI> subjects = new ArrayList<>();
+            Optional<Type> type;
+            try {
+                if (mongoTypeStorage == null) {
+                    mongoTypeStorage = new MongoTypeStorage(mongo, ryaInstanceName);
+                }
+                type = mongoTypeStorage.get(firstType);
+            } catch (final TypeStorageException e) {
+                throw new EntityStorageException("Unable to get entity type: " + firstType, e);
+            }
+            if (type.isPresent()) {
+                // Grab the subjects for all the types we found matching "firstType"
+                final ConvertingCursor<TypedEntity> cursor = search(Optional.empty(), type.get(), Collections.emptySet());
```

Comment:
Unfortunately, we can't use the properties to query for potential duplicates.  That would bring back Entities with EXACT matches of the properties.  We want to include Entities that would be NEAR matches (based on the tolerance).  So, if we have a tolerance of 1% for longitudes we'd want to consider 99° to be the same as an Entity with 100° which means we couldn't query for it based off its property value of 100°.

I kind of think Entities should be judged as a whole and not as part of their components.  If we have 2 completely different Joe Smiths it's possible that they're supposed to have 2 Employee TypedEntities and only one of them has a Person TypedEntity.  If we try to create the one that only has a Employee TypedEntity and we say it's a duplicate of the other Joe Smith's Employee TypedEntity (due to setting high tolerance values) then it won't get created.  But if we saw that they had different TypedEntities associated with them then we'd consider them to not be duplicates and both Joe Smith's Entities would be created.